Move to community membership model closer to kubernetes one

serathius opened the pull request:
Based on https://github.com/kubernetes/community/blob/master/community-membership.md

Changes:
* Extracted contributor membership to separate file
* Provide more detailed requirements for each role. Base maintainers on kubernetes subproject owners.
* Introduction of member role


cc @ahrtr @ptabor @spzala @mitake

chaochn47 commented on Documentation/contributor-guide/contributor-membership.md:
Typo: `Members`

serathius replied:
Done

chaochn47 commented on Documentation/contributor-guide/contributor-membership.md:
How to define `substantial`? Is it tagged with `priority/important`?

ahrtr replied:
I think at least it isn't PR something like fixing typo, but not necessary that it must be `priority/important`.

chaochn47 commented on Documentation/contributor-guide/contributor-membership.md:
Just curious, how does this governance rules apply to `etcd-io/bbolt`?

ahrtr replied:
As the maintainers discussed, we will add separate Maintainer file in bbolt, probably in raft as well. cc @tbg 

I think the roles enumerated in this doc should be applied to all projects under etcd-io, at least including bbolt and raft. WDYT? @hexfusion @mitake @ptabor @serathius @spzala @tbg

serathius commented on Documentation/contributor-guide/contributor-membership.md:
Would love to but was not able to google what permissions are needed to start workflows. Maybe triage is enough, but let me know if you find anything.

serathius replied:
https://docs.github.com/en/actions/managing-workflow-runs/re-running-workflows-and-jobs
```
Who can use this feature
People with write permissions to a repository can re-run workflows in the repository.
```

serathius replied:
It would require a write permissions, which would allow to also merge PRs. https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/repository-roles-for-an-organization
That's too much even for reviewers

serathius replied:
We would need k8s-ci-robot to separate rerunning tests from write permissions :(.

I considered it some time ago as it apparently can be run from github workflows, however it might be too much burden as we don't have expertise. We can continue the discussion as a followup.

serathius replied:
Yea, before becoming maintainer I also re-pushed appended commit with force. Would be a great perk, but impossible with current github setup.

ahrtr commented on Documentation/contributor-guide/contributor-membership.md:
```suggestion
- Sponsored by two active maintainers.
```

ahrtr commented on Documentation/contributor-guide/contributor-membership.md:
I think the retiring process should be applied to all roles, not just Maintainers.

ahrtr replied:
We may also want to clearly request all roles (Members, Reviewers and Maintainers) to notify the community when they are going to be inactive for more than 1 week, e.g. on vacation without computer.

ptabor replied:
I wouldn't require this from Members definitely. I'm not sure about reviewers but probably not. 
That make sense for maintainers, as it's likely blocking for others.

serathius replied:
Agree with @ptabor, I'm also not super pressed to clarify process of removing members/reviewers, as we first need to nominate them :P. Let's do it in followup.

ahrtr commented on Documentation/contributor-guide/contributor-membership.md:
No strong opinion on this, but be good to discuss it separately.

spzala replied:
I think that would be overkill for the project. The current structure seems good enough, but yes agree that it can be discussed separately in the future.

serathius replied:
In my opinion using OWNERS are great for a big project to scale reviews, however I don't think etcd project will or should grow further. We are also already splitting the codebase into separate repos which allows us to avoid per directory owners. 

I think that adding OWNERS is not aligned with project direction which is to simplify and reduce size of codebase.

ptabor replied:
I'm seeing OWNERS as a way to give 'approvers' privileges to less sensitive parts of the code-base like tests, tools or docs that might be an additional engagement tool. 
I agree with direction to limit scope of the code-base, but even if we cut 'proxy' out of `etcd-io/etcd` repo and put it as `etcd-io/etcd-proxy`, than it might make sense (if anyone is willing to) to have separate approvers/writers for this repo (this means we would have sub-project OWNERS, but not implemented using the OWNERs file yet).

ptabor commented on Documentation/contributor-guide/contributor-membership.md:
Not sure if this line make sense if we require **one** active maintainer. 

In the header you had "2 reviewers" and than it made sense.

ahrtr replied:
+1

serathius replied:
Done

ptabor commented on Documentation/contributor-guide/contributor-membership.md:
reviewed "OR MERGED" -> Members cannot 'merge' and this suggests when side by side to "reviewed" working on other's PRs. 

"reviewed or/and contributed at least 20 substantial PRs"

ptabor reviewed (approved):
Thank you for working on this. 
Few details, but overall LGTM.

spzala reviewed (approved):
Thanks @serathius  
LGTM with couple of small points: 
- address @ptabor comment on clarifying Member requirements mismatch "2 reviewers in the table vs one maintainer in the details" 
- we should provide a link to this doc in the CONTRIBUTING.doc to encourage new contributors to check membership ladder

ahrtr reviewed (approved):
LGTM with minor comments as summarized by @spzala + a couple of typos

serathius commented on Documentation/contributor-guide/contributor-membership.md:
```diff
@@ -0,0 +1,151 @@
+
+### Requirements
+
+- Have made multiple contributions to the project or community.  Contribution may include, but is not limited to:
```
Didn't want to touch removing members at this time. For maintainers we already have a process described below with 12 months deadline. For now we want to onboard new members, let's tackle removal as a follow up.

ptabor replied:
I have also this thought. "Triage" permissions is a nothing that put's organisation on risk. 
If we had auto-test running -> it's something that could be used for abuse... but I wouldn't solve not existing yet problem.

serathius commented on Documentation/contributor-guide/contributor-membership.md:
This is a great followup. I would not want to enforce this without previous discussion and preparation. Don't want to cut access to people without proper notification. There are also etcd org members for other subprojects that would be affected.

serathius commented on GOVERNANCE.md:
```diff
@@ -5,84 +5,16 @@
-Reviewers can be removed by a supermajority of the  maintainers or can resign by notifying
-the maintainers.
+please see [contributor guide] guide.
 
```
Added

serathius commented:
Addressed all comments. With supermajority of maintainers (4/6) approving the PR we are ready to merge, still I will leave the PR open for couple of days to ensure everyone has time to read it again. Please take a look and feel free to file issues with follow up ideas.

ptabor reviewed (approved):
1. LGTM.

I hate to say that, but the only concern I have is that with the requirement of 2 active maintainers, we might risk getting 'deadlock' at some moment that might require
"breaking" the governance rules in such situation. 

We might:
  - postpone solving the problem when it (hopefully not) happens...
  - declare lazy consensus, i.e. lack of disagreement of any maintainer for 2 weeks is like an agreement. 
  - delegate solving deadlocks to e.g. k8s-leads (if they happen).

serathius commented:
I think we should follow https://github.com/etcd-io/etcd/blob/main/GOVERNANCE.md#conflict-resolution in case of deadlock

serathius commented:
Everyone involved left a thumbs up on https://github.com/etcd-io/etcd/pull/15593#issuecomment-1494461938 so I assume they have familiarize with the latest version. Merging.
